Bug while using dogleg optimizer

I have created an unit test (`examples/SFMExample_SmartFactor_bug.cpp`) to illustrate the issue and can be found in my [fork](https://github.com/kvmanohar22/gtsam/tree/dogleg_bug)

In the discussion in [issue:367](https://bitbucket.org/gtborg/gtsam/issues/367/isam2-with-smart-factors-null-ptr-segfault), this bug was resolved but seems like it is popping up again.

## Description

This error seems to pop up when using Dogleg optimizer as part of ISAM2. Following is the reason for this error.

Assume `graph_` represents the existing graph (passed previously to `gtsam::ISAM2`) that has smart factors. Suppose a smart factor was already in the previous graph and we update the smart factor with a new measurement from a pose (this pose was not part of previous graph).

Assuming that the following condition holds true ([line](https://github.com/borglab/gtsam/blob/b4eb0c233d683eaf47511ee9b062dc3c822484df/gtsam/nonlinear/ISAM2.cpp#L424))
```cpp
  // Update delta if we need it to check relinearization later
  if (update.relinarizationNeeded(update_count_))
    updateDelta(updateParams.forceFullSolve);
```
the error can be reproduced. Specifically, the error arises from the following [line](https://github.com/borglab/gtsam/blob/b4eb0c233d683eaf47511ee9b062dc3c822484df/gtsam/nonlinear/ISAM2.cpp#L746) which tries to compute the error of the previous graph but without updating `theta_`

I have tried to simulate this in the unit test.

```cpp
nonlinearFactors_.error(theta_),
```

## Steps to reproduce

Just run the new unit test that I have added (in my fork) using

```bash
make SFMExample_SmartFactor_bug.run
```
which results in the following error

```bash
kv@kv:~/slam/gtsam/build$ make SFMExample_SmartFactor_bug.run
[ 33%] Built target metis
[100%] Built target gtsam
[100%] Building CXX object examples/CMakeFiles/SFMExample_SmartFactor_bug.dir/SFMExample_SmartFactor_bug.cpp.o
[100%] Linking CXX executable SFMExample_SmartFactor_bug
[100%] Built target SFMExample_SmartFactor_bug
Initial Estimates:
Values with 3 values:
Value 0: (N5gtsam5Pose3E) R:
[
 	 0.208654,  0.073369, -0.975233;
    	 0.949229, -0.255245,  0.183888;
    	-0.235432, -0.964089, -0.122902
  ]
[29.8, 0.05, 0.1]';

Value 1: (N5gtsam5Pose3E) R:
[
 	-0.523665,  0.232365, -0.819623;
    	 0.818747, -0.128606, -0.559566;
    	-0.235432, -0.964089, -0.122902
  ]
[21.0364, 21.1071, 0.1]';

Value 2: (N5gtsam5Pose3E) R:
[
 	-0.949229,  0.255245, -0.183888;
    	 0.208654,  0.073369, -0.975233;
    	-0.235432, -0.964089, -0.122902
  ]
[-0.05, 29.8, 0.1]';

Results:
Values with 3 values:
Value 0: (N5gtsam5Pose3E) R:
[
 	-0.0253294,  0.0208786,  -0.999461;
    	  0.998264, -0.0526448, -0.0263988;
    	-0.0531676,  -0.998395, -0.0195089
  ]
[30.0013, -0.000248155, 0.00250433]';

Value 1: (N5gtsam5Pose3E) R:
[
 	  -0.709983,   0.0133575,   -0.704092;
    	   0.704144, -0.00108604,   -0.710056;
    	 -0.0102493,    -0.99991, -0.00863457
  ]
[21.2138, 21.2135, 0.00132106]';

Value 2: (N5gtsam5Pose3E) R:
[
 	 -0.999205, -0.0254096,  0.0307042;
    	-0.0320221,  0.0531939,  -0.998071;
    	 0.0237273,  -0.998261, -0.0539653
  ]
[-2.03514, 30.1004, 1.98536]';

terminate called after throwing an instance of 'gtsam::ValuesKeyDoesNotExist'
  what():  Attempting to at the key "3", which does not exist in the Values.
Aborted (core dumped)
examples/CMakeFiles/SFMExample_SmartFactor_bug.run.dir/build.make:57: recipe for target 'examples/CMakeFiles/SFMExample_SmartFactor_bug.run' failed
make[3]: *** [examples/CMakeFiles/SFMExample_SmartFactor_bug.run] Error 134
CMakeFiles/Makefile2:2660: recipe for target 'examples/CMakeFiles/SFMExample_SmartFactor_bug.run.dir/all' failed
make[2]: *** [examples/CMakeFiles/SFMExample_SmartFactor_bug.run.dir/all] Error 2
CMakeFiles/Makefile2:2667: recipe for target 'examples/CMakeFiles/SFMExample_SmartFactor_bug.run.dir/rule' failed
make[1]: *** [examples/CMakeFiles/SFMExample_SmartFactor_bug.run.dir/rule] Error 2
Makefile:884: recipe for target 'SFMExample_SmartFactor_bug.run' failed
make: *** [SFMExample_SmartFactor_bug.run] Error 2

```

## Expected behavior
The error should not have popped up

## Environment
OS: Ubuntu 16.04
Language: C++, custom build

## Additional information
This bug was discussed previously in the following discussion. [link](https://bitbucket.org/gtborg/gtsam/issues/367/isam2-with-smart-factors-null-ptr-segfault), specifically comment by Sean Bowman.

--- a/gtsam/geometry/Rot2.h
+++ b/gtsam/geometry/Rot2.h
@@ -68,7 +68,7 @@ namespace gtsam {
       return fromAngle(theta * degree);
     }
 
-    /// Named constructor from cos(theta),sin(theta) pair, will *not* normalize!
+    /// Named constructor from cos(theta),sin(theta) pair
     static Rot2 fromCosSin(double c, double s);
 
     /**

--- a/gtsam/nonlinear/ISAM2.cpp
+++ b/gtsam/nonlinear/ISAM2.cpp
@@ -424,6 +424,11 @@ ISAM2Result ISAM2::update(const NonlinearFactorGraph& newFactors,
   ISAM2Result result(params_.enableDetailedResults);
   UpdateImpl update(params_, updateParams);
 
+  // Initialize any new variables \Theta_{new} and add
+  // \Theta:=\Theta\cup\Theta_{new}.
+  // Needed before delta update if using Dogleg optimizer.
+  addVariables(newTheta, result.details());
+
   // Update delta if we need it to check relinearization later
   if (update.relinarizationNeeded(update_count_))
     updateDelta(updateParams.forceFullSolve);
@@ -435,9 +440,7 @@ ISAM2Result ISAM2::update(const NonlinearFactorGraph& newFactors,
   update.computeUnusedKeys(newFactors, variableIndex_,
                            result.keysWithRemovedFactors, &result.unusedKeys);
 
-  // 2. Initialize any new variables \Theta_{new} and add
-  // \Theta:=\Theta\cup\Theta_{new}.
-  addVariables(newTheta, result.details());
+  // 2. Compute new error to check for relinearization
   if (params_.evaluateNonlinearError)
     update.error(nonlinearFactors_, calculateEstimate(), &result.errorBefore);
 
@@ -731,6 +734,7 @@ void ISAM2::updateDelta(bool forceFullSolve) const {
                                       effectiveWildfireThreshold, &delta_);
     deltaReplacedMask_.clear();
     gttoc(Wildfire_update);
+
   } else if (std::holds_alternative<ISAM2DoglegParams>(params_.optimizationParams)) {
     // If using Dogleg, do a Dogleg step
     const ISAM2DoglegParams& doglegParams =
@@ -769,9 +773,8 @@ void ISAM2::updateDelta(bool forceFullSolve) const {
     gttic(Copy_dx_d);
     // Update Delta and linear step
     doglegDelta_ = doglegResult.delta;
-    delta_ =
-        doglegResult
-            .dx_d;  // Copy the VectorValues containing with the linear solution
+    // Copy the VectorValues containing with the linear solution
+    delta_ = doglegResult.dx_d;
     gttoc(Copy_dx_d);
   } else {
     throw std::runtime_error("iSAM2: unknown ISAM2Params type");

--- a/tests/testDoglegOptimizer.cpp
+++ b/tests/testDoglegOptimizer.cpp
@@ -184,6 +184,14 @@ TEST(DoglegOptimizer, Constraint) {
 }
 
 /* ************************************************************************* */
+/**
+ * Test created to fix issue in ISAM2 when using the DogLegOptimizer.
+ * Originally reported by kvmanohar22 in issue #301
+ * https://github.com/borglab/gtsam/issues/301
+ *
+ * This test is based on a script provided by kvmanohar22
+ * to help reproduce the issue.
+ */
 TEST(DogLegOptimizer, VariableUpdate) {
   // Make the typename short so it looks much cleaner
   typedef SmartProjectionPoseFactor<Cal3_S2> SmartFactor;

--- a/tests/testDoglegOptimizer.cpp
+++ b/tests/testDoglegOptimizer.cpp
@@ -24,10 +24,9 @@
 #include <gtsam/nonlinear/DoglegOptimizerImpl.h>
 #include <gtsam/nonlinear/NonlinearEquality.h>
 #include <gtsam/slam/BetweenFactor.h>
-#include <gtsam/inference/Symbol.h>
-#include <gtsam/linear/JacobianFactor.h>
-#include <gtsam/linear/GaussianBayesTree.h>
-#include <gtsam/base/numericalDerivative.h>
+#include <gtsam/nonlinear/ISAM2.h>
+#include <gtsam/slam/SmartProjectionPoseFactor.h>
+#include "examples/SFMdata.h"
 
 #include <functional>
 
@@ -36,7 +35,6 @@ using namespace gtsam;
 
 // Convenience for named keys
 using symbol_shorthand::X;
-using symbol_shorthand::L;
 
 /* ************************************************************************* */
 TEST(DoglegOptimizer, ComputeBlend) {
@@ -183,6 +181,120 @@ TEST(DoglegOptimizer, Constraint) {
 #else
   CHECK_EXCEPTION(optimizer2.optimize(), std::invalid_argument);
 #endif
+}
+
+/* ************************************************************************* */
+TEST(DogLegOptimizer, VariableUpdate) {
+  // Make the typename short so it looks much cleaner
+  typedef SmartProjectionPoseFactor<Cal3_S2> SmartFactor;
+
+  // create a typedef to the camera type
+  typedef PinholePose<Cal3_S2> Camera;
+  // Define the camera calibration parameters
+  Cal3_S2::shared_ptr K(new Cal3_S2(50.0, 50.0, 0.0, 50.0, 50.0));
+
+  // Define the camera observation noise model
+  noiseModel::Isotropic::shared_ptr measurementNoise =
+      noiseModel::Isotropic::Sigma(2, 1.0);  // one pixel in u and v
+
+  // Create the set of ground-truth landmarks and poses
+  vector<Point3> points = createPoints();
+  vector<Pose3> poses = createPoses();
+
+  // Create a factor graph
+  NonlinearFactorGraph graph;
+
+  ISAM2DoglegParams doglegparams = ISAM2DoglegParams();
+  doglegparams.verbose = false;
+  ISAM2Params isam2_params;
+  isam2_params.evaluateNonlinearError = true;
+  isam2_params.relinearizeThreshold = 0.0;
+  isam2_params.enableRelinearization = true;
+  isam2_params.optimizationParams = doglegparams;
+  isam2_params.relinearizeSkip = 1;
+  ISAM2 isam2(isam2_params);
+
+  // Simulated measurements from each camera pose, adding them to the factor
+  // graph
+  unordered_map<int, SmartFactor::shared_ptr> smart_factors;
+  for (size_t j = 0; j < points.size(); ++j) {
+    // every landmark represent a single landmark, we use shared pointer to init
+    // the factor, and then insert measurements.
+    SmartFactor::shared_ptr smartfactor(new SmartFactor(measurementNoise, K));
+
+    for (size_t i = 0; i < poses.size(); ++i) {
+      // generate the 2D measurement
+      Camera camera(poses[i], K);
+      Point2 measurement = camera.project(points[j]);
+
+      // call add() function to add measurement into a single factor, here we
+      // need to add:
+      //    1. the 2D measurement
+      //    2. the corresponding camera's key
+      //    3. camera noise model
+      //    4. camera calibration
+
+      // add only first 3 measurements and update the later measurements
+      // incrementally
+      if (i < 3) smartfactor->add(measurement, i);
+    }
+
+    // insert the smart factor in the graph
+    smart_factors[j] = smartfactor;
+    graph.push_back(smartfactor);
+  }
+
+  // Add a prior on pose x0. This indirectly specifies where the origin is.
+  // 30cm std on x,y,z 0.1 rad on roll,pitch,yaw
+  noiseModel::Diagonal::shared_ptr noise = noiseModel::Diagonal::Sigmas(
+      (Vector(6) << Vector3::Constant(0.3), Vector3::Constant(0.1)).finished());
+  graph.emplace_shared<PriorFactor<Pose3> >(0, poses[0], noise);
+
+  // Because the structure-from-motion problem has a scale ambiguity, the
+  // problem is still under-constrained. Here we add a prior on the second pose
+  // x1, so this will fix the scale by indicating the distance between x0 and
+  // x1. Because these two are fixed, the rest of the poses will be also be
+  // fixed.
+  graph.emplace_shared<PriorFactor<Pose3> >(1, poses[1],
+                                            noise);  // add directly to graph
+
+  // Create the initial estimate to the solution
+  // Intentionally initialize the variables off from the ground truth
+  Values initialEstimate;
+  Pose3 delta(Rot3::Rodrigues(-0.1, 0.2, 0.25), Point3(0.05, -0.10, 0.20));
+  for (size_t i = 0; i < 3; ++i)
+    initialEstimate.insert(i, poses[i].compose(delta));
+  // initialEstimate.print("Initial Estimates:\n");
+
+  // Optimize the graph and print results
+  isam2.update(graph, initialEstimate);
+  Values result = isam2.calculateEstimate();
+  // result.print("Results:\n");
+
+  // we add new measurements from this pose
+  size_t pose_idx = 3;
+
+  // Now update existing smart factors with new observations
+  for (size_t j = 0; j < points.size(); ++j) {
+    SmartFactor::shared_ptr smartfactor = smart_factors[j];
+
+    // add the 4th measurement
+    Camera camera(poses[pose_idx], K);
+    Point2 measurement = camera.project(points[j]);
+    smartfactor->add(measurement, pose_idx);
+  }
+
+  graph.resize(0);
+  initialEstimate.clear();
+
+  // update initial estimate for the new pose
+  initialEstimate.insert(pose_idx, poses[pose_idx].compose(delta));
+
+  // this should break the system
+  isam2.update(graph, initialEstimate);
+  result = isam2.calculateEstimate();
+  EXPECT(std::find(result.keys().begin(), result.keys().end(), pose_idx) !=
+         result.keys().end());
 }
 
 /* ************************************************************************* */

